[StanleyNeoh] iP

StanleyNeoh opened the pull request:
# Duke
> "Productivity is never an accident. It is always the result of a commitment to excellence, intelligent planning, and focused effort." -- Paul J. Meyer ([source](https://www.goodreads.com/quotes/1133520-productivity-is-never-an-accident-it-is-always-the-result#:~:text=It%20is%20always%20the%20result,intelligent%20planning%20and%20focused%20effort.%E2%80%9D))

Duke is a **text-based** task management application that enables you to manage your task with just a keyboard.

Some of the features include:
* Text-based with minimalist GUI
* Intuitive and easy to learn
* Fast to use

How do you get started? All you need to do is,
1. Clone the repository into your machine
1. Navigate into the repository
1. Run `java -jar build/libs/duke.jar` and watch it go!

You may run the following to get started
```
git clone git@github.com:nus-cs2103-AY2223S2/ip.git
cd ip
java -jar build/libs/duke.jar
```

Commit:
Add Gradle support

StanleyNeoh committed:
Implemented Level-1: Greet, Echo, Exit

StanleyNeoh committed:
Implemented Level-2: Add, List

StanleyNeoh committed:
Implemented Level 3. Mark as Done

StanleyNeoh committed:
Implemented Level-4: ToDo, Event, Deadline

StanleyNeoh committed:
Added spaces

StanleyNeoh committed:
Added A-TextUiTesting

StanleyNeoh committed:
Implemented Level-5: Handle Errors

StanleyNeoh committed:
Implemented Level-6: Delete

StanleyNeoh committed:
Implement Level-7: Save

StanleyNeoh committed:
Implement Level-8: Dates and Times

StanleyNeoh committed:
Merge branch 'branch-Level-7'

StanleyNeoh committed:
Merge branch 'branch-Level-8'

StanleyNeoh committed:
Bug-fix after merging

StanleyNeoh committed:
Implement A-MoreOOP, A-Package

StanleyNeoh committed:
Merge branch 'add-gradle-support'

StanleyNeoh committed:
Implement A-Gradle

StanleyNeoh committed:
Implement A-JUnit

StanleyNeoh committed:
Improve OOP of Serializer

StanleyNeoh committed:
Make main class into a package

StanleyNeoh committed:
Implement A-CodingStandard

StanleyNeoh committed:
Implement A-JavaDoc

StanleyNeoh committed:
Merge tag 'A-CodingStandard'

StanleyNeoh committed:
Merge branch 'A-JavaDoc'

StanleyNeoh committed:
Delete Deserializer.java

StanleyNeoh committed:
Fix merge styling issues

StanleyNeoh committed:
Implement Level-9: Find

StanleyNeoh committed:
Merge branch 'branch-Level-9'

StanleyNeoh committed:
Fix styling issues after merging

StanleyNeoh committed:
Fix data path for A-Jar

StanleyNeoh committed:
Implement Level-10: GUI

StanleyNeoh committed:
Merge branch 'branch-Level-10'

PROGRAMMERHAO commented on src/main/java/duke/storage/serializer/TaskSerializer.java:
```diff
@@ -0,0 +1,39 @@
+     * @return Task defined by instance
+     * @throws DukeException
+     */
+    public Task createTask() throws DukeException {
```
I like how you use verb-like method name here

PROGRAMMERHAO commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,71 @@
+        this.completed = false;
+    }
+
+    public boolean matches(String pattern) {
```
Maybe here matches can be converted to something like canMatch?

PROGRAMMERHAO commented on src/main/java/duke/storage/serializer/Serializer.java:
```diff
@@ -0,0 +1,50 @@
+        }
+    }
+
+    public void add(String key, Object value) {
```
Maybe a more descriptive name here can help the reader understand your code better? Such as addToMap?

PROGRAMMERHAO commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,146 @@
+     *
+     * @param chunks sequence of strings to be printed.
+     */
+    public void respond(String... chunks) {
```
Perhaps something like respondUser could be an alternative here?

PROGRAMMERHAO commented on src/main/java/duke/tasklist/TaskList.java:
```diff
@@ -0,0 +1,128 @@
+        return t;
+    }
+
+    public int length() {
```
Could getLength() work here?

PROGRAMMERHAO reviewed (commented):
The overall code quality is very good! With >95% of code following good naming convention, I really like how you name boolean attributes and methods. Just some minor details to consider as given in the comment. LGTM!

Lava-Iris commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,98 @@
+    @Override
+    public void start(Stage stage) {
+        //Initialise
+        scrollPane = new ScrollPane();
```
Maybe having separate methods for initialise, styling etc. would make the code easier to read?

Lava-Iris commented on src/main/java/duke/core/Core.java:
```diff
@@ -0,0 +1,102 @@
+    }
+
+    private void handleCommand(String cmd) throws DukeException {
+        if (cmd.matches("^bye$")) {
```
Perhaps switch case can be used here? Switch is usually faster and easier to read. It might be possible with the java.util.regex package and a hash table, though it might be more complicated.

Lava-Iris commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,79 @@
+ * Represents a Task with a deadline.
+ */
+public class Deadline extends Task {
+    private static final String ICON = "D";
```
I like how you abstracted out all the constants here.

Lava-Iris commented on src/test/java/duke/tasklist/TaskListTest.java:
```diff
@@ -0,0 +1,81 @@
+
+public class TaskListTest {
+    @Test
+    public void addAndGet() {
```
Maybe addAndGetTask would be a bit more descriptive?

Lava-Iris commented on src/test/java/duke/tasklist/TaskListTest.java:
```diff
@@ -0,0 +1,81 @@
+    }
+
+    @Test
+    public void delete() {
```
Perhaps deleteTask instead of delete would be more descriptive here?

Lava-Iris reviewed (commented):
Hi Stanley! 

I really like your program structure and flow. I couldn't find any major problems in your code, and the documentation is well done. 

All the best for future increments!

StanleyNeoh committed:
Implement A-Assertions

Various section of the code makes assumptions of the state of the program.

Add assert statements to act as null-checks and verify the
state of the program at various key areas.

Assert statements are disabled by default and will not affect the performance of the program.
However, it will indicate to the programmer the requirements of the state at that point.

StanleyNeoh committed:
Implement A-CodeQuality

Duke class initialises, styles the GUI and manages the user input which leads to
a very lengthy class that is difficult to maintain.

Extract the main gui portion into another class ChatPane which handles the
initialising and styling of the GUI.

Pulling it into its own class follows better OOP practise as encapsulation and abstraction
are utilised to make the GUI more maintainable. ChatPane inheriting the AnchorPane class allows
us to utilise a ChatPane instance like an AnchorPane.

StanleyNeoh committed:
Implement A-Streams

In the TaskList class, the matches method utlises a for-loop to iterate
through tasks and filter them. A stream.filter method could be used here
instead.

Replace for-loop implementation with stream.filter implementation.

Utilising streams better convey the intention of the method as compared
to the for-loop implementation.

StanleyNeoh committed:
Merge pull request #1 from StanleyNeoh/branch-A-Assertions

Implement A-Assertions

StanleyNeoh committed:
Merge pull request #2 from StanleyNeoh/branch-A-CodeQuality

Implement A-CodeQuality

StanleyNeoh committed:
Merge pull request #3 from StanleyNeoh/branch-A-Streams

Implement A-Streams

StanleyNeoh committed:
Implement A-CI

StanleyNeoh committed:
Implement B-DoAfterTasks

StanleyNeoh committed:
Fix no file found exception

StanleyNeoh committed:
Fix naming issue of Task.java completed attribute

StanleyNeoh committed:
Change profile pictures

StanleyNeoh committed:
Write userguide

StanleyNeoh committed:
Fix image cropping

StanleyNeoh committed:
Use FXML

StanleyNeoh committed:
Merge pull request #5 from StanleyNeoh/branch-fmxl

Use FXML
